src/components/useFetch.tsx

```diff
@@ -0,0 +1,64 @@
+import { useState, useEffect } from 'react'
+
+type Post = {
+    userId: number;
+    id: number;
+    title: string;
+    body: string;
+}
+type FetchResult = {
+    data: Array<Post>;
+    loading: boolean;
+    error: string;
+    onSuccess: (data: Array<Post>) => void;
+    onError: (error: string) => void;
+    refetch: () => void;
+}
+
+const useFetch = (url: string): FetchResult => {
+
+    const [data, setData] = useState<Array<Post>>([])
+    const [loading, setLoading] = useState(true)
+    const [error, setError] = useState('')
+
+    const fetchData = async (url: string) => {
+        try {
+            const resp = await fetch(url);
+
+            if (!resp.ok) {
+                throw new Error(resp.status.toString());
+            }
+
+            const responseData = await resp.json();
+            setLoading(false);
+            setData(responseData);
+            setError('');
+            onSuccess(responseData);
+        } catch (error: any) {
+            setData([])
+            setLoading(false);
+            const errorMessage = error instanceof Error ? error.message : 'Unknown error occurred';
+            setError(errorMessage);
+            onError(errorMessage);
+        }
+    }
+    const onSuccess = (data: Array<Post>) => {
```

Comment:
`onSuccess` should be prop, as you sometimes need to do something after request is success.

Reply:
Fixed.